File: imgur.py
import requests
import argparse
import json
from bs4 import BeautifulSoup
import re
from validators import url as check_url



class ImgurParser(object):

    def __init__(self, min_height=0, min_width=0, 
                         max_height=-1, max_width=-1, 
                         height=-1, width=-1):
        self.min_height = min_height
        self.min_width = min_width
        self.max_height = max_height
        self.max_width = max_width
        self.height = height
        self.width = width

        self.routes = [
          "^https?://imgur.com/gallery/([^/]+)$",
          "^https?://imgur.com/r/[^/]+/([^/]+)$",
          "^https?://imgur.com/([^/]+)$",
          "^https?://imgur.com/t/[^/]+/([^/]+)$",
          "^https?://imgur.com/a/([^/]+)$",
          "^https?://imgur.com/([^/]+)$",
        ]
    

    def get_images(self, url):

        rel = ""

        if not check_url(url):
            return []

        for route in self.routes:
            m = re.search(route, url)
            if m:
                rel = m.group(1)

        if not rel:
            return []

        gallery_base="http://imgur.com/ajaxalbums/getimages/{}/hit.json"

        response = requests.get(url)
        html = response.text

        soup = BeautifulSoup(html, "lxml")
        images=[]
      

        gallery_url = gallery_base.format(rel)
        gallery_base.format(rel)
        json_res = requests.get(gallery_url)
        gallery = json.loads(json_res.text)
        
        if gallery['data']:
            if 'images' in gallery['data']:
                images=gallery['data']['images']
            images=[ 'http://i.imgur.com/'+x['hash']+x['ext'] for x in images ]
        else:
            images = []
            for img in soup.find_all("div", {"class":"post-image-container"}):
                soup=BeautifulSoup(str(img), "lxml")
                image=soup.find("img")
                if image:
                    images.append("http:"+image["src"])
                elif img:
                    #slow last resort
                    images+=get_images("http://i.imgur.com/" + img['id'])

        return images 

File: src/imgur.rs
use regex::Regex;
use url::Url;
use reqwest;
use kuchiki::traits::*;
use kuchiki;

use parser::{SiteParser,Error};
use parser::Error::*;
use parser::RouteError::*;


#[derive(Debug,Clone)]
pub struct ImgurParser {
    routes: Vec<Regex>,
}

#[derive(Debug, Deserialize)]
struct GalleryResponse {
    data: Gallery
}

#[derive(Debug, Deserialize)]
pub struct Gallery {
    count: usize,
    pub images: Vec<Image>,
}

#[derive(Debug, Deserialize)]
pub struct Image {
    pub hash:  String,
    pub title: String,
    pub description: Option<String>,
    pub has_sound: bool,
    pub width: usize,
    pub height: usize,
    pub size: usize,
    pub ext: String,
    pub animated: bool,
    pub prefer_video: bool,
    pub looping: bool,
    pub datetime: String,
}

impl Image {
    pub fn url(&self) -> String {
        format!("http://i.imgur.com/{}{}", self.hash, self.ext)
    }
}

impl SiteParser for ImgurParser {
    fn parse(&self, uri: &str) -> Vec<String> {

        let (i,re) = self.routes.iter().enumerate().find(|&(_,x)| {
                x.is_match(uri)
            }).unwrap();

        if i == 6 {
            vec![uri.to_string()]
        } else {
            let id = re.captures(uri).map(|m| m.get(1).unwrap())
                .unwrap();

            self.image_list(uri, id.as_str()).unwrap_or(vec![])
        }
    }

    fn is_valid(&self, uri: &str) -> bool {
        let routes = self.routes.iter().find(|x| {
                x.is_match(uri)
            }).map_or(false, |_| true);
        let url = Url::parse(uri).is_ok();

        routes && url
    }
}

impl ImgurParser {

    pub fn new() -> Self {
        let routes: Vec<_> = vec![
            Regex::new(r"^https?://imgur\.com/gallery/([^/]+)$").unwrap(),
            Regex::new(r"^https?://imgur\.com/r/[^/]+/([^/]+)$").unwrap(),
            Regex::new(r"^https?://imgur\.com/([^/]+)$").unwrap(),
            Regex::new(r"^https?://imgur\.com/t/[^/]+/([^/]+)$").unwrap(),
            Regex::new(r"^https?://imgur\.com/a/([^/]+)$").unwrap(),
            Regex::new(r"^https?://imgur\.com/([^/]+)$").unwrap(),
            Regex::new(r"^https?://i\.imgur\.com/([^/]+)$").unwrap(),
        ];

        ImgurParser {
            routes
        }
    }

    fn image_list(&self, uri: &str, id: &str) -> Result<Vec<String>, Error> {
        if let Ok(gallery) = self.gallery(uri, id) {
            Ok(gallery.images.iter().map(|x| x.url()).collect())
        } else if let Ok(images) = self.images(uri) {
            Ok(images)
        } else {
            Err(Route(InvalidGallery))
        }
    }

    fn images(&self, uri: &str) -> Result<Vec<String>, Error> {
        let html = reqwest::get(uri)?.text()?;
        let post_container = "div.post-image-container";

        let document = kuchiki::parse_html().one(html);

        let images = document.select(post_container).unwrap()
            .filter_map(|div| {
                div.as_node().select_first("img").ok()
            }).flat_map(|img| {
                self.source_from_element(&*img)
            }).collect();

        Ok(images)
    }

    fn source_from_element(&self, img: &kuchiki::ElementData) -> Vec<String> {
        if let Some(src) = img.attributes.borrow().get("src") {
            vec![format!("http:{}", src)]
        } else if let Some(id) = img.attributes.borrow().get("id") {
            self.parse(&format!("http://i.imgur.com/{}", id))
        } else {
            vec![]
        }
    }

    fn gallery(&self, uri: &str, id: &str) -> Result<Gallery, Error> {

        if let Err(_) = Url::parse(uri) {
            return Err(Route(InvalidUrl));
        }

        let ajax_request = &format!(
            "http://imgur.com/ajaxalbums/getimages/{}/hit.json",
            id,
        );

        let gallery: GalleryResponse = reqwest::get(ajax_request)?.json()?;
        Ok(gallery.data)
    }
}

